test/IDE/complete_single_expression_return.swift

```diff
@@ -128,6 +128,55 @@ struct TestSingleExprClosureGlobal {
 // TestSingleExprClosureGlobal-DAG: Decl[InstanceMethod]/CurrNominal: void()[#Void#];
 }
 
+struct TestSingleExprClosureBinding {
+  func void() -> Void {}
+  func str() -> String { return "" }
+  func int() -> Int { return 0 }
+
+  func test() -> Int {
+    let fn = {
+      self.#^TestSingleExprClosureBinding^#
+    }
+    return fn()
+  }
+// Void is always valid in an implicit single expr closure.
+// TestSingleExprClosureBinding-DAG: Decl[InstanceMethod]/CurrNominal: str()[#String#];
+// TestSingleExprClosureBinding-DAG: Decl[InstanceMethod]/CurrNominal: int()[#Int#];
+// TestSingleExprClosureBinding-DAG: Decl[InstanceMethod]/CurrNominal: void()[#Void#];
+}
+
+struct TestExplicitSingleExprClosureBinding {
+  func void() -> Void {}
+  func str() -> String { return "" }
+  func int() -> Int { return 0 }
+
+  func test() {
+    let fn = {
+      return self.#^TestExplicitSingleExprClosureBinding^#
+    }
+  }
+// FIXME: Because we have an explicit return, and no expected type, we shouldn't suggest Void.
```

Comment:
Actually, I think suggesting values that return `Void` after `return` makes sense if the closure has a `Void` return type. For example `return callback()` is a common pattern for early returns out of closures and as a shorthand of `callback(); return`

So, I would just remove the FIXME

Reply:
Right, but in this case there's no contextual type. If there was an explicit `Void` contextual type, we'd allow it

Reply:
We could carve out an exception for `CTP_ReturnStmt` if we really wanted this behavior though

Reply:
I think what I’m trying to say is that I don’t really care about this case because you can argue either way and I would just remove the FIXME. But if you want to keep it, that’s good with me as well.

Reply:
Going to keep it for now since this is going to change anyway with my constraint system patch. I should also note that not suggesting Void is also the behavior we currently have for multi-statement closure returns. I don't really feel too strongly about it either way (and am happy to carve out an exception for returns if we want that), but I think we should at least be consistent there

Reply:
Changing in https://github.com/apple/swift/pull/71272